core: Tidy contributor onboarding, fix typos.

GirlBossRush opened the pull request:
<!--
👋 Hi there! Welcome.

Please check the Contributing guidelines: https://docs.goauthentik.io/docs/developer-docs/#how-can-i-contribute
-->

## Details

<!--
Explain what this PR changes, what the rationale behind the change is, if any new requirements are introduced or any breaking changes caused by this PR.

Ideally also link an Issue for context that this PR will close using `closes #`
-->

This PR contains a _subset_ of fixes discovered when progressing through the authentik contributor guide. Highlights include:

- Merging of disparate code spell dictionaries
- Optional Poetry env for hygiene commands (e.g. lint, build, etc)
- Fix for stale links, typos.


---

## Checklist

-   [x] Local tests pass (`ak test authentik/`)
-   [x] The code has been formatted (`make lint-fix`)

If an API change has been made

-   [ ] The API schema has been updated (`make gen-build`)

If changes to the frontend have been made

-   [ ] The code has been formatted (`make web`)

If applicable

-   [ ] The documentation has been updated
-   [ ] The documentation has been formatted (`make website`)

rissson commented on pyproject.toml:
```diff
@@ -4,6 +4,32 @@ version = "2024.12.2"
+    "schema.yaml",
+    "schema.json",
+    "go.sum",
+    "cmd",
+    "internal",
+    "web/src",
+    "locale",
+    "**/web/xliff/**",
+    "./website/blog/**",
+    "./website/docs/**",
+    "./website/integrations/**",
+    "./website/src/**",
```
```suggestion
    "locale",
    "**/web/xliff/**",
    "./website/src/**",
```

Many of those we definitely don't want to skip. It's nice to catch spelling mistakes in the code. You can also remove `authentik` from that list.

On top of that, I would say that this might cause more problems down the line with generated files from the tools we use (like `__pycache__`, `gen-go-api`, `gen-ts-api`, `unittest.xml` to name a few`). I would rather go with the previous approach of including what needs to be included and leaving the rest alone.

rissson replied:
But then again, we were missing a few things like `scripts`. If codespell could ignore the same things as .gitignore, plus a few generated committed things that'd be great, if we can reuse the .gitignore file, let's not duplicate that list

GirlBossRush commented on website/package.json:
```diff
@@ -1,12 +1,17 @@
     "license": "MIT",
     "scripts": {
         "aws-cfn": "cd docs/install-config/install/aws && cdk synth > template.yaml && ./fix_template.py",
-        "build": "cp ../docker-compose.yml static/docker-compose.yml && cp ../schema.yml static/schema.yaml && docusaurus gen-api-docs all && cross-env NODE_OPTIONS='--max_old_space_size=65536' docusaurus build",
-        "build-bundled": "cp ../schema.yml static/schema.yaml && docusaurus gen-api-docs all && cross-env NODE_OPTIONS='--max_old_space_size=65536' docusaurus build",
+        "build": "run-s build:docker-compose build-bundled",
+        "build-bundled": "run-s build:schema build:api-docs build:docusaurus",
+        "build:api-docs": "docusaurus gen-api-docs all",
+        "build:docker-compose": "cp ../docker-compose.yml static/docker-compose.yml",
+        "build:docusaurus": "cross-env NODE_OPTIONS='--max_old_space_size=65536' docusaurus build",
+        "build:schema": "cp ../schema.yml static/schema.yml",
```
@rissson Found the source of the mystery `.yaml` file 😊

I've tidied up the build sub-scripts to aid in readability.

GirlBossRush commented on Makefile:
```diff
@@ -55,23 +41,22 @@ test-docker:  ## Run all tests in a docker-compose
-	coverage run manage.py test --keepdb authentik
-	coverage html
-	coverage report
+	poetry run coverage run manage.py test --keepdb authentik
```
Hi @rissson!

Thanks for reviewing my PR. I'm open to either approach, but it may help to keep the prefix.

When running `make test` without a Poetry shell activated, `make` doesn't know where to find the `coverage` command and emits an error without a hint at the solution.

```sh
❯ make test
coverage run manage.py test --keepdb authentik
make: coverage: No such file or directory
make: *** [test] Error 1
```

The virtual environment instructions appear in the [full-stack setup docs](https://docs.goauthentik.io/docs/developer-docs/setup/full-dev-environment#backend-setup), but my concern is that contributors who start with the [front-end](https://docs.goauthentik.io/docs/developer-docs/setup/frontend-dev-environment) or [docs](https://docs.goauthentik.io/docs/developer-docs/setup/website-dev-environment) environments might miss the Python requirements when switching contexts.

Since ci-main.yml uses the poetry prefix consistently, would it make sense to keep it for now? We could look at removing it from both places as a separate task if you think that would be valuable.

Thanks for your feedback!

rissson replied:
So if you add `poetry run` invocations to the Makefile, then it makes sense to remove it from the CI.

Both approaches (having it in CI or having it in the Makefile) are valid, but I think we should only do one of them. I agree that having it in the Makefile is easier for new contributors

rissson commented on Makefile:
```diff
@@ -55,23 +41,21 @@ test-docker:  ## Run all tests in a docker-compose
+	poetry run codespell -w
 
 lint: ## Lint the python and golang sources
-	bandit -r $(PY_SOURCES) -x web/node_modules -x tests/wdio/node_modules -x website/node_modules
-	golangci-lint run -v
```
Not sure why those got removed

rissson commented on Makefile:
```diff
@@ -55,23 +41,21 @@ test-docker:  ## Run all tests in a docker-compose
 
 core-install:
-	poetry install
+	poetry run bandit -c pyproject.toml -r $(PY_SOURCES)
```
```suggestion
	poetry install
```

rissson commented on Makefile:
```diff
@@ -274,16 +258,17 @@ ci--meta-debug:
 
 ci-bandit: ci--meta-debug
-	bandit -r $(PY_SOURCES)
+	poetry run bandit -r $(PY_SOURCES)
+	golangci-lint run -v internal/...
```
```suggestion
	poetry run bandit -r $(PY_SOURCES)
```

rissson commented on manage.py:
```diff
@@ -4,11 +4,11 @@
 from defusedxml import defuse_stdlib
 from django.utils.autoreload import DJANGO_AUTORELOAD_ENV
 
+from authentik.lib.config import CONFIG
```
```suggestion
```

rissson commented on pyproject.toml:
```diff
@@ -12,7 +32,7 @@ exclude = 'node_modules'
 line-length = 100
 target-version = "py312"
-exclude = ["**/migrations/**", "**/node_modules/**"]
+exclude = ["web/tools/**", "**/migrations/**", "**/node_modules/**"]
```
Why are we excluding those scripts from being linted and formatted?

rissson commented on Makefile:
```diff
@@ -4,7 +4,7 @@
 UID = $(shell id -u)
 GID = $(shell id -g)
-NPM_VERSION = $(shell python -m scripts.npm_version)
+NPM_VERSION = $(shell python -m scripts.generate_semver)
```
We could also do this

```suggestion
NPM_VERSION = $(shell poetry run scripts/generate_semver.py)
```

rissson commented on pyproject.toml:
```diff
@@ -4,6 +4,26 @@ version = "2024.12.2"
+    "schema.yml",
+    "./blueprints/schema.json",
+    "go.sum",
+    "cmd",
+    "internal",
+    "web/src",
```
```suggestion
```

These are source code directories that we want to ignore as well

tanberry reviewed (commented):
I can't say one way or the other on the code, but thanks for catching the misspellings!!

GirlBossRush committed:
core: Tidy contributor onboarding.

- Fixes typos.
- Fixes stale links.
- Tidies Makefile so that Poetry env is optional for hygiene commands.
- Remove mismatched YAML naming.
- Uses shebang on Python scripts.
- Document semver usage.
- Redirect OpenAPI schema.

rissson commented on website/docusaurus.config.ts:
```diff
@@ -134,7 +134,7 @@ module.exports = async function (): Promise<Config> {
                     config: {
                         authentik: {
-                            specPath: "static/schema.yaml",
+                            specPath: "static/schema.yml",
```
We should add a redirect for this in the netlify.toml

rissson reviewed (changes requested):
One more little thing I didn't catch previously, mb

GirlBossRush commented:
@BeryJu Rebased to latest main and ready for review 😁

BeryJu reviewed (commented):
Need to verify if we need a redirect for schema.yml on the website repo
Need to check if we need to update any docs that directly run the `ak` executable to add a `poetry run` prefix

GirlBossRush commented on website/docusaurus.config.ts:
```diff
@@ -152,7 +152,7 @@ const createConfig = (): Config => {
                     config: {
                         authentik: {
-                            specPath: "static/schema.yaml",
+                            specPath: "static/schema.yml",
```
@BeryJu https://github.com/goauthentik/website/pull/216
